docs/Components.md

```diff
@@ -0,0 +1,137 @@
+    import ExecutionContext.Implicits.global
+    val srv = Await.result(server.init, Duration.Inf)
+    srv.join()
+  }
+}
+```
+
+## `Component`
+
+The `Component` type can be summarized as an enhanced version of Scala's built-in `lazy val` construct.
+It is associated with a lazy-evaluated expression that creates and initializes the component (often a 
+simple constructor invocation, but it may be an arbitrary piece of code). 
+
+Wrapping component definitions into `Component` instances relieves the programmer from manually figuring out 
+the initialization order of multiple interdependent components. Note that simply using `lazy val`s would already
+do that job, but `Component` does this better because:
+
+* in case of failures, exceptions are wrapped in a way that lets you track the initialization path that
+  leads to the faulty component (i.e. a stack trace but nicer to read than the raw one)
+* dependency cycles are detected before any initialization is done - a nice error message is displayed that lets
+  you quickly see where the cycle is
+  
+`Component` instances are aware of their source position. If they are assigned to a `val` or `def`, the name of
+this `val` or `def` is interpreted as the component name. This name may be later useful for debugging. 
+Here's an example of a dependency cycle error that uses component names and source positions:
+
+```
+Exception in thread "main" com.avsystem.commons.di.DependencyCycleException: component dependency cycle detected:
+  server(MyApp.scala:13) ->
+  service(MyApp.scala:16) ->
+  database(MyApp.scala:19) ->
+  server(MyApp.scala:13)
+```
+
+## Entry points
+
+In order to start our application in the previous example, we explicitly requested initialization of the 
+`server` component. This way we assumed `server` to be an _entry point_ of our application - the primary 
+service that exposes core system functionality (e.g. an HTTP server).
+All the other components are initialized only as direct or indirect dependencies of this toplevel entry point.
+
+If your application consists of multiple such entry points, e.g. an UI server, REST API server, FTP server, etc.
+then it is recommended to create a single toplevel component whose purpose is solely to aggregate all the 
+"real" entry points into one.
+
+```scala
+import com.avsystem.commons.di._
+
+class UiServer
+class ApiServer
+class FtpServer
+
+class FullApplication(ui: UiServer, api: ApiServer, ftp: FtpServer)
+
+object MyApp extends Components {
+  val ui: Component[UiServer] = component(new UiServer)
+  val api: Component[ApiServer] = component(new ApiServer)
+  val ftp: Component[FtpServer] = component(new FtpServer)
+  
+  val fullApplication: Component[FullApplication] =
+    component(new FullApplication(ui.ref, api.ref, ftp.ref))
+}
+```
+
+This way your application always has a single entry point component which makes it easier to navigate
+through code and understand its structure (knowing where to start).
```

Comment:
That depends on what you mean by "separate entry points" - does this imply two applications running completely independently or are some components shared (component instances, in runtime)?

Reply:
I had completely independent apps in mind (not even single shared dependency between them).

I guess when I have some shared components I should wrap them with one `single toplevel component`

Reply:
For separate applications, you can have separate objects extending `Components` or possibly separate instances of the same `Components` implementation (two instances of the same app). You can still share component _definitions_ via inheritance. If you need to share actual component _instances_ then you can use composition rather than inheritance.